### ISSUE TITLE
Proposal for performance improvement - Spatial and edge encoding -

### ISSUE THREAD

**Issue by xtibau** [leffff/graphormer-pyg#32]
The spatial encoding can be easily optimized by using pytorch indexing. 
One can define a function that stores the `phi(i,j)` values and use that in b as in this example:

```python
# By using a matrx P that stores the value of phi(i, j) there is no need of a loop
# and we can still propagate the gradients in b and A.

import torch

b = torch.tensor([10., 20., 30.], requires_grad=True)
P = torch.tensor([[1, 2, 1],
                 [3, 1, 2]])
A = torch.randn(2, 3, requires_grad=True)

result = A + b[P-1]  # This substitutes the loop in the forward.
```

The matrix P only needs to be computed once. 

I hope this helps.

**Comment by xtibau** [comment]
Edge encoding could follow a similar approach. Here is an example. The AI has implemented the establishment of the path tensor inside the layer during initialization, but this could have been done before. This way, the loop is only computed once, and everything is matricial operations.

**Comment by leffff** [comment]
Hi, @xtibau !
You can create a branch, I will test it and merge it!
I would be glad If you could optimize the code and commit to the implementation!

**Comment by jq-ding** [comment]
Actually, the `edge_paths` and `edge_attr` that used for EdgeEncoding won't be changed in attention calculation, so maybe `c = self.edge_encoding(x, edge_attr, edge_paths)` could be moved before `self.layers` in` forward` of main model, so it'd be compute once in the whole feedforward process.

**Comment by leffff** [comment]
I believe, it should be computed in each layer. Correct me if I'm wrong.

**Comment by jq-ding** [comment]
> I believe, it should be computed in each layer. Correct me if I'm wrong.
oh, you're right, new parameters need to be init every time when edge encoding is called

**Comment by xtibau** [comment]
Hi both, 

> Actually, the `edge_paths` and `edge_attr` that used for EdgeEncoding won't be changed in attention calculation, so maybe `c = self.edge_encoding(x, edge_attr, edge_paths)` could be moved before `self.layers` in` forward` of main model, so it'd be compute once in the whole feedforward process.

This conversation is interesting. The paper says: 
> To better encode edge features into attention layers, we propose a new edge encoding method in Graphormer. The attention mechanism needs to estimate correlations for each node pair (vi, vj), and we believe the edges connecting them should be considered in the correlation as in [34, 51]. For each ordered node pair (vi, vj), we find (one of) the shortest path SPij = (e1, e2, ..., eN) from vi to vj, and compute an average of the dot-products of the edge feature and a learnable embedding along the path.

For that, I understand that the SP can be computed before or in the init of the layer, and the same for the edge features (which are constant, I understand). So the only thing that is required to have gradients and calculated each time the product of the embedding (shared across layers, maybe from here comes the confusions about being reusable) with the features. 

Do you have the same understanding as I?

**Comment by leffff** [comment]
cb644583802e1e76240247ba95735dc8fc6678e1
